jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/PreloadFormHandler.java

```diff
@@ -0,0 +1,87 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.ee10.servlet;
+
+import java.util.concurrent.CompletableFuture;
+
+import jakarta.servlet.ServletRequest;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.MimeTypes;
+import org.eclipse.jetty.server.FormFields;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.util.Callback;
+
+/**
+ * Handler to asynchronously preload & parse {@link MimeTypes.Type#FORM_ENCODED} and
+ * {@link MimeTypes.Type#MULTIPART_FORM_DATA} content prior to invoking the {@link ServletHandler},
+ * which can then consume them with blocking APIs but without blocking.
+ * @see FormFields#from(Request)
+ * @see ServletMultiPartFormData#from(ServletRequest)
+ */
+public class PreloadFormHandler extends Handler.Wrapper
+{
+    public PreloadFormHandler()
+    {
+        this(null);
+    }
+
+    public PreloadFormHandler(Handler handler)
+    {
+        super(handler);
+    }
+
+    @Override
+    public boolean handle(Request request, org.eclipse.jetty.server.Response response, Callback callback) throws Exception
+    {
+        String contentType = request.getHeaders().get(HttpHeader.CONTENT_TYPE);
+        if (contentType == null)
+            return super.handle(request, response, callback);
+
+        MimeTypes.Type mimeType = MimeTypes.getBaseType(contentType);
+        if (mimeType == null)
+            return super.handle(request, response, callback);
+
+        CompletableFuture<?> future =  switch (mimeType)
+        {
+            case FORM_ENCODED -> FormFields.from(request);
+            case MULTIPART_FORM_DATA -> ServletMultiPartFormData.from(Request.as(request, ServletContextRequest.class).getServletApiRequest(), contentType);
+            default -> null;
+        };
+
+        if (future == null)
+            return super.handle(request, response, callback);
+
+        if (future.isDone())
+        {
+            if (!super.handle(request, response, callback))
+                callback.failed(new IllegalStateException("Not Handled"));
+            return true;
+        }
+
+        future.whenComplete((result, failure) ->
+        {
+            try
```

Comment:
There must be a check on `failure` to not call super if we are already failed.

In case of success, we want to store the parts as an attribute, no?

Reply:
result or failure are handled by ServletApiRequest#getParts.  I'll comment.